infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -29,33 +30,39 @@
  * The layout server serves layouts, which are used by clients to find the
  * Corfu infrastructure.
  *
- * <p>For replication and high availability, the layout server implements a
- * basic Paxos protocol. The layout server functions as a Paxos acceptor,
+ * <p>For replication and high availability, the layout server implements
+ * Basic Paxos with NACKs. The layout server functions as a Paxos acceptor,
  * and accepts proposals from clients consisting of a rank and desired
- * layout. The protocol consists of three rounds:
+ * value (layout). The protocol consists of three rounds:
  *
  * <p>1)   Prepare(rank) - Clients first contact each server with a rank.
  * If the server responds with ACK, the server promises not to
  * accept any requests with a rank lower than the given rank.
- * If the server responds with LAYOUT_PREPARE_REJECT, the server
- * informs the client of the current high rank and the request is
+ * ACK consists of a triple (proposed rank, accepted rank, accepted value).
+ * Both accepted rank and the value are optional in case nothing has been accepted.
+ * If the server responds with NACK (LAYOUT_PREPARE_REJECT), the server
+ * informs the client of the current highest proposed rank and the request is
  * rejected.
  *
- * <p>2)   Propose(rank,layout) - Clients then contact each server with
- * the previously prepared rank and the desired layout. If no other
+ * <p>2)   Propose(rank, layout) - Clients then contact each server with
+ * the previously prepared rank and the desired value (layout). If no other
  * client has sent a prepare with a higher rank, the layout is
  * persisted, and the server begins serving that layout to other
  * clients. If the server responds with LAYOUT_PROPOSE_REJECT,
  * either another client has sent a prepare with a higher rank,
  * or this was a propose of a previously accepted rank.
  *
- * <p>3)   Committed(rank, layout) - Clients then send a hint to each layout
- * server that a new rank has been accepted by a quorum of
- * servers.
+ * <p>3)   Committed(rank, layout) - Clients then send a hint to each acceptor
+ * (layout server) that a new proposal has been accepted by a quorum of
+ * servers. Adding phase three to Classic Paxos serves an important purpose
+ * that may not be immediately apparent, namely that the liveness conditions
+ * are no longer necessary for progress. With this variant, Classic Paxos can make
+ * progress provided either a majority of acceptors are up or at least one acceptor
+ * who has been notified of decision is up. As a result, the proposer may return a
+ * decided value after communicating with just one acceptor.
  *
  * <p>Created by mwei on 12/8/15.
  */
-//TODO Finer grained synchronization needed for this class.
 //TODO Need a janitor to cleanup old phases data and to fill up holes in layout history.
```

Comment:
should we get rid of this TODO, it looks like it's useless

Reply:
I think we do not want to remove old registers, but before removing this TODO, we should probably have a discussion on this.